lib/gpu/src/context.rs

```diff
@@ -0,0 +1,408 @@
+        unsafe {
+            self.device.vk_device().reset_fences(&[self.vk_fence])?;
+        }
+
+        Ok(())
+    }
+
+    fn init_command_buffer(&mut self) -> GpuResult<()> {
+        if self.vk_command_buffer != vk::CommandBuffer::null() {
+            return Err(GpuError::Other(
+                "Vulkan command buffer was already created".to_string(),
+            ));
+        }
+
+        // Create new command buffer from pool.
+        let command_buffer_allocate_info = vk::CommandBufferAllocateInfo::default()
+            .command_pool(self.vk_command_pool)
+            .level(vk::CommandBufferLevel::PRIMARY)
+            .command_buffer_count(1);
+        self.vk_command_buffer = unsafe {
+            self.device
+                .vk_device()
+                .allocate_command_buffers(&command_buffer_allocate_info)?[0]
+        };
+
+        let command_buffer_begin_info =
+            vk::CommandBufferBeginInfo::default().flags(vk::CommandBufferUsageFlags::default());
+        //.inheritance_info(..);
+
+        let begin_result = unsafe {
+            self.device
+                .vk_device()
+                .begin_command_buffer(self.vk_command_buffer, &command_buffer_begin_info)
+        };
+
+        // If command buffer creation failed, destroy created command buffer and return error.
+        if let Err(e) = begin_result {
+            self.destroy_command_buffer();
+            return Err(GpuError::from(e));
+        }
+
+        Ok(())
+    }
+
+    fn destroy_command_buffer(&mut self) {
+        if self.vk_command_buffer != vk::CommandBuffer::null() {
+            unsafe {
+                self.device
+                    .vk_device()
+                    .free_command_buffers(self.vk_command_pool, &[self.vk_command_buffer]);
+            }
+            self.vk_command_buffer = vk::CommandBuffer::null();
+        }
+        self.resources.clear();
+    }
+}
+
+impl Drop for Context {
+    fn drop(&mut self) {
+        let wait_result = self.wait_finish(DROP_TIMEOUT);
+        match wait_result {
+            Err(GpuError::Timeout) => {
+                // Timeout reached, resources are still in use.
+                // Vulkan API cannot stop GPU execution.
+                // This sutiation may appear if shader has infinite loop, etc.
+                // There is no good way to handle this error.
+                // So just log it and ignore resources deallocation.
+                // This approach may cause memory leaks and used gpu kernels,
+                // but it's better than potential segfault.
+                log::error!("Failed to wait for GPU context to finish");
+
+                // Error was logged, do memory leak to keep the gpu running.
+                let resources = self.resources.clone();
+                self.resources.clear();
+                for resource in resources.into_iter() {
+                    // !!!!!!!!!
+                    std::mem::forget(resource);
```

Comment:
I need additional eyes here. If a shader has an infinity loop, I allow to run it and leak all gpu related resources with proper log error.